### PR TITLE
Guarantee delivery for fullscreen and restart events from shidur to sndman, sdiout, audioout apps. Lib e2e + test. 

### PR DESCRIPTION
**Pull request by bbfsdev**

### CODE REVIEW

File: src/apps/VirtualApp/VirtualChat.js [hunk 1]
```diff
@@ -13,6 +19,8 @@ class VirtualChat extends Component {
     support_msgs: [],
     room_chat: true,
     from: null,
+    // TODO: Share one delivery manager for all required components?
+    gdm: new GuaranteeDeliveryManager(MAX_DELAY, RETRY_DELAY, INTERVALS_DELAY),
```
Comment (edoshor) [inline src/apps/VirtualApp/VirtualChat.js:23]: Empty constructor should use sane defaults. 
No need to pass these around

File: src/apps/VirtualApp/VirtualChat.js [hunk 2]
```diff
@@ -92,29 +100,43 @@ class VirtualChat extends Component {
...
+          this.setState({ support_msgs, from });
+          if (this.props.visible) {
+            this.scrollToBottom();
+          } else {
+            notifyMe('Shidur', message.text, true);
+            this.setState({ room_chat: false });
+            this.props.onNewMsg(true);
+          }
+        }).catch((error) => {
+          console.log('Failed receiving private message', error);
```
Comment (edoshor) [inline src/apps/VirtualApp/VirtualChat.js:118]: use console.error for errors

File: src/apps/VirtualApp/VirtualChat.js [hunk 3]
```diff
@@ -150,13 +172,36 @@ class VirtualChat extends Component {
...
+    const onSuccess = (msg) => {
+      const { room_chat, support_msgs } = this.state;
+      Janus.log(':: Message sent ::');
+      this.setState({ input_value: '' });
+      if (!room_chat) {
+        support_msgs.push(msg);
+        this.setState({ support_msgs });
+      }
+    }
+    this.state.gdm.sendGuaranteeMessage(msg, (m) => this.sendChatMessage_(m)).then(() => {
```
Comment (edoshor) [inline src/apps/VirtualApp/VirtualChat.js:192]: `(m) => this.sendChatMessage_(m)` could be simply `this.sendChatMessage_`

File: src/shared/GuaranteeDelivery.js [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+      }
+    }
+  }
+
+  // Send message that guarantee to arrive within MAX_DELAY or notify on timeout.
+  // @param {!Object} message
+  // @param {!Array<string>} toAck List of userIds that are expected to ack. If empty will expect any non-self user to ack.
+  // @param {function(!Object)} sendMessage
+  // @return {!Promise}  <====  Resolve for success, reject for failure.
+  sendGuaranteeMessage(message, toAck, sendMessage) {
```
Comment (edoshor) [inline src/shared/GuaranteeDelivery.js:72]: I'd use shorter names: `send`, `accept` and `checkAck`.
The words 'Guarantee' and 'Message' are implicit for users of this class

File: src/apps/VirtualApp/VirtualChat.js (hunk 1 above)
Comment (bbfsdev) [reply src/apps/VirtualApp/VirtualChat.js]: Fixed by default values.

File: src/apps/VirtualApp/VirtualChat.js (hunk 3 above)
Comment (bbfsdev) [reply src/apps/VirtualApp/VirtualChat.js]: Removed this usage example, not really needed.

**Pull request closed without merging**